evap/staff/importers/enrollment.py

```diff
@@ -0,0 +1,764 @@
+            name_en, merge_hindrances = key
+            self.importer_log.add_error(
+                format_html(
+                    _(
+                        "{location}: Course {course_name} already exists in this semester, but the courses can not be merged for the following reasons:{reasons}"
+                    ),
+                    location=location_string,
+                    course_name=f'"{name_en}"',
+                    reasons=format_html_join("", "<br /> - {}", ([msg] for msg in merge_hindrances)),
+                ),
+                category=ImporterLogEntry.Category.COURSE,
+            )
+
+        for name_de, location_string in self.name_de_collision_tracker.aggregated_keys_and_location_strings():
+            self.importer_log.add_error(
+                _(
+                    '{location}: Course "{course_name}" (DE) already exists in this semester with a different english name.'
+                ).format(
+                    location=location_string,
+                    course_name=name_de,
+                ),
+                category=ImporterLogEntry.Category.COURSE,
+            )
+
+        for name_en, location_string in self.name_en_collision_tracker.aggregated_keys_and_location_strings():
+            self.importer_log.add_error(
+                _(
+                    '{location}: Course "{course_name}" (EN) already exists in this semester with a different german name.'
+                ).format(
+                    location=location_string,
+                    course_name=name_en,
+                ),
+                category=ImporterLogEntry.Category.COURSE,
+            )
+
+        for key, location_string in self.name_de_mismatch_tracker.aggregated_keys_and_location_strings():
+            self.importer_log.add_error(
+                _(
+                    '{location}: The German name for course "{course_name}" is already used for another course in the import file.'
+                ).format(
+                    location=location_string,
+                    course_name=key[0],
+                ),
+                category=ImporterLogEntry.Category.COURSE,
+            )
+
+
+class CourseDataMismatchChecker(Checker):
+    """Assert CourseData is consistent between rows"""
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.course_data_by_name_en: Dict[str, CourseData] = {}
+        self.tracker: FirstLocationAndCountTracker = FirstLocationAndCountTracker()
+
+    def check_course_data(self, course_data: CourseData, location: ExcelFileLocation):
+        if not all_fields_valid(course_data):
+            return
+
+        stored = self.course_data_by_name_en.setdefault(course_data.name_en, course_data)
+
+        # degrees will be merged together
+        if course_data.equals_when_ignoring_degrees(stored):
+            return
+
+        self.tracker.add_location_for_key(location, course_data.name_en)
```

Comment:
I slightly favor the other way, because we have 2 conditions here that allow a course to "pass".

With that schema (multiple passes, a single failure), I like it more when we use early returns. See https://github.com/e-valuation/EvaP/blob/main/evap/evaluation/models.py#L1611-L1620